### ISSUE TITLE
Unable to bake lightmap with CSG

### ISSUE THREAD

**Issue by theraot** [godotengine/godot#47766]
**Godot version:**
Godot 3.2.4-beta6 to Godot 3.3-rc8

**OS/device including version:**
Windows 10
NVIDIA GeForce GTX 750 Ti
GLES2/GLES3

**Issue description:**

BakedLightmap does not work with CSG nodes. As far as I can tell, this worked up to beta 5.

As far as I can tell the CSG node does not have a UV2 layer. *And I cannot use "Create Outline Mesh" + "Unwrap UV2 for Lightmap/AO" with it (The Mesh option in the toolbar does not appear for CSG nodes).*

**Steps to reproduce:**

- Create new Godot project (GLES2 or GLES3, does not matter).
- Create new Spatial scene.
- Save it, any name.
- Add CSGBox (or any other CSG node) to the scene. *Observe there is no Mesh icon in the bar when the CSG node is selected.*
- Set `use_in_baked_light = true` of the CSG node.
- Add BakedLightmap node (make sure the CGS node is inside).
- Click BakeLightmap, select any path, click "Save".

Result: Message appears "No meshes to bake. Make sure they contain UV2 channel and `Bake Light` flag is on."

Expected Result: The light is baked (which is no light, since we didn't add any, and thus the CSG should appear dark).

**Minimal reproduction project:**
[BakeMeUp.zip](https://github.com/godotengine/godot/files/6289322/BakeMeUp.zip)

**Comment by Calinou** [comment]
> Add CSGBox (or any other CSG node) to the scene. Observe there is no Mesh icon in the bar when the CSG node is selected.

This is intended, since CSGBox doesn't inherit from MeshInstance.

To be fair, we'd need https://github.com/godotengine/godot-proposals/issues/182 to be implemented first to make this a reliable option. Baking UV2 on a CSG node whose geometry can change on a whim isn't necessarily a good idea…

**Comment by theraot** [comment]
> Baking UV2 on a CSG node whose geometry can change on a whim isn't necessarily a good idea…

For context, I wanted to test with CSG nodes because they would give a quick cycle of reshaping them, baking lights, see changes, repeat. The intention was to try to find issues with baked lights.

**Comment by kalbfled** [comment]
I authored the StackExchange question @theraot references above.  I attempted the same basic test he describes, except I used MeshInstances with CubeMesh and PlaneMesh as the mesh data.  I couldn't bake light with that setup because Godot refuses to unwrap meshes that are not instances of ArrayMesh.  This doesn't make sense to me when these meshes already have a mapping in uv1.  Maybe this is a candidate for an enhancement--allow baking with PrimitiveShape instances.  I worked around by creating a plane+cube scene in Blender and transferring it into Godot as .glb.

**Comment by theraot** [comment]
@kalbfled I ended up working around that by using "Create Outline Mesh" which would create ArrayMesh from what I had, and then "Unwrap UV2 for Lightmap/AO". I arrived to that searching online. This is also the motivation to mention that that is not an option with CSG on this issue.

**Comment by theraot** [comment]
Is this UV2 not the same UV2? Because this seems to have UV2, but still can't bake:

![image](https://user-images.githubusercontent.com/1212509/114876501-018b7680-9dc4-11eb-8c76-b77d50d9da51.png)

**Comment by Zireael07** [comment]
If you are referring to the uv2 property of the detail section of the material - nope, that's only telling the material how to scale UV2 IF it exists. And CSG doesn't have UV2.

**Comment by theraot** [comment]
@Zireael07 Well, it scaled something. The material is showing a texture applied according to the Uv 1 property section, combined to a texture applied according to the Uv 2 property section, combined according to the Detail property section. I also tried converting that to shader code, and it used UV2. So, if anything, that is confusing.

Was it a decision to remove UV2 from CSG? Or was it that baking didn't use UV2 before? (As I stated in OP, baking worked in beta5 and older, including stable).

**Comment by Calinou** [comment]
> Was it a decision to remove UV2 from CSG? Or was it that baking didn't use UV2 before? (As I stated in OP, baking worked in beta5 and older, including stable).

Baking always used UV2, and *baking* UV2 was never supported on CSG nodes.

**Comment by theraot** [comment]
@Calinou 

Alright, what I'm understanding is that CSG does not have UV2, and that baking working before was as accident. In that order of ideas, I propose a solution for this issue: remove the baking light properties from CSG nodes (`use_in_baking_light`, `generate_lightmap` and `lightmap_scale`).

Also, somebody may want to see into the naming of Uv2. The CSG has something called like that, that works.

**Comment by Zireael07** [comment]
Yep, I have to admit the naming in your specific situation is confusing indeed. Regarding the materials being actually displayed, it's even more confusing.... some sort of a silent fallback to UV1 because UV2 doesn't exist?

**Comment by Calinou** [comment]
> In that order of ideas, I propose a solution for this issue: remove the baking light properties from CSG nodes (use_in_baking_light, generate_lightmap and lightmap_scale).

Those properties are available in any GeometryInstance node, and CSG nodes inherit from GeometryInstance. I don't think it makes sense to hide them there, or even if it's technically possible.

**Comment by theraot** [comment]
@Calinou I suppose we have a wontfix then.

Just to be clear, this was not supported: https://youtu.be/vEFYSpxnfdE

**Comment by ee0pdt** [comment]
> Those properties are available in any GeometryInstance node, and CSG nodes inherit from GeometryInstance. I don't think it makes sense to hide them there, or even if it's technically possible.

I could understand it not being technically possible - but surely it makes a lot of sense to hide them if it is possible?

I just wasted a fair amount of time trying to understand why CSG meshes were not baking! Could we at least have a warning added?

**Comment by Calinou** [comment]
Is there any example of the editor hiding a node property that's inherited from its parent (and not hidden there)?

**Comment by ee0pdt** [comment]
@Calinou I'm not aware of any, but that is not the only option we could explore here. I'm just suggesting we find a way to inform users that these properties have no effect for a given node.

The closest thing I can think of is the warning on a Container node:

![image](https://user-images.githubusercontent.com/6094917/135298312-9646205c-96b6-408e-af24-eedd5096cb76.png)

**Comment by Calinou** [comment]
@ee0pdt I removed my old comment because it was incorrect. It should be possible to hide the properties using `_validate_property()` as done here: https://github.com/godotengine/godot/blob/27d4e2f09fe76e06e6b4751025acc79d993c61e5/modules/csg/csg_shape.cpp#L571-L579

**Comment by Zireael07** [comment]
Speaking of xatlas, IIRC it doesn't like the built in meshes (e.g. cubemesh to use 4.0 terminology) either?

**Comment by Calinou** [comment]
> Speaking of xatlas, IIRC it doesn't like the built in meshes (e.g. cubemesh to use 4.0 terminology) either?

It's not xatlas' fault – it's because Godot does not have a way to store custom UV2 for primitive meshes.

**Comment by akien-mga** [comment]
Is this still reproducible in 3.5 beta 5 or later?

**Comment by Calinou** [comment]
> Is this still reproducible in 3.5 beta 5 or later?

Yes, as far as I know, nothing changed on the subject.

**Comment by theraot** [comment]
Tested in Godot 3.6. It is still the same situation.

**Comment by Calinou** [comment]
> Tested in Godot 3.6. It is still the same situation.

In 4.x, this is addressed by having the ability to convert CSG nodes to MeshInstance3D, which you can then bake UV2 on using the **Mesh** menu at the top of the 3D editor viewport.

I don't think this will be addressed in 3.x unless someone reimplements conversion to mesh there, which is unlikely given most of the development focus is on 4.x now.

**Comment by ItsFoxy87** [comment]
There is a plugin that exports CSGCombiners into an OBJ mesh, which can then be reimported as a mesh instance. However, it still seems to lack UV2 regardless because Godot still refuses to accept it as a candidate for lightmap baking.
